From f54a78ea3c5e618be015d2f29c3c1243f8b532fe Mon Sep 17 00:00:00 2001
From: Wojciech Bartczak <wbartczak@marvell.com>
Date: Thu, 14 Jan 2021 14:46:19 -0800
Subject: [PATCH 1223/1921] driver: mmc: Adds device tree entries to control
 eMMC input timings

The change bases on existing mechanism to give user a control over
input timings for eMMC bus. This should ensure thet user is able
to adjust timings for the board without changing the source code for mmc
driver.

Following entries are available for different modes:
- marvell,cmd-in-hs200-dly
- marvell,data-in-hs200-dly
- marvell,cmd-in-hs400-dly
- marvell,data-in-hs400-dly
- marvell,cmd-in-hs-sdr-dly
- marvell,data-in-hs-sdr-dly
- marvell,cmd-in-hs-ddr-dly
- marvell,data-in-hs-ddr-dly
- marvell,cmd-in-legacy-dly
- marvell,data-in-legacy-dly

The values entered using device tree are subject to further
modification by tuning mechanism for MMC HS 200 and MMC HS 400 modes.

Signed-off-by: Wojciech Bartczak <wbartczak@marvell.com>
Change-Id: I35f98c98a78f9c1523d098c77b08928d22e3ba96
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/44775
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/mmc/host/cavium.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/drivers/mmc/host/cavium.c b/drivers/mmc/host/cavium.c
index f5767399bb57..6ca5fd586f80 100644
--- a/drivers/mmc/host/cavium.c
+++ b/drivers/mmc/host/cavium.c
@@ -2201,6 +2201,27 @@ static int cvm_mmc_of_parse(struct device *dev, struct cvm_mmc_slot *slot)
 			     &slot->cmd_out_taps_dly[MMC_TIMING_LEGACY]);
 	of_property_read_u32(node, "marvell,data-out-legacy-dly",
 			     &slot->data_out_taps_dly[MMC_TIMING_LEGACY]);
+	/* Modify the input timings using user inputs */
+	of_property_read_u32(node, "marvell,cmd-in-hs200-dly",
+			     &slot->cmd_in_taps_dly[MMC_TIMING_MMC_HS200]);
+	of_property_read_u32(node, "marvell,data-in-hs200-dly",
+			     &slot->data_in_taps_dly[MMC_TIMING_MMC_HS200]);
+	of_property_read_u32(node, "marvell,cmd-in-hs400-dly",
+			     &slot->cmd_in_taps_dly[MMC_TIMING_MMC_HS400]);
+	of_property_read_u32(node, "marvell,data-in-hs400-dly",
+			     &slot->data_in_taps_dly[MMC_TIMING_MMC_HS400]);
+	of_property_read_u32(node, "marvell,cmd-in-hs-sdr-dly",
+			     &slot->cmd_in_taps_dly[MMC_TIMING_MMC_HS]);
+	of_property_read_u32(node, "marvell,data-in-hs-sdr-dly",
+			     &slot->data_in_taps_dly[MMC_TIMING_MMC_HS]);
+	of_property_read_u32(node, "marvell,cmd-in-hs-ddr-dly",
+			     &slot->cmd_in_taps_dly[MMC_TIMING_MMC_DDR52]);
+	of_property_read_u32(node, "marvell,data-in-hs-ddr-dly",
+			     &slot->data_in_taps_dly[MMC_TIMING_MMC_DDR52]);
+	of_property_read_u32(node, "marvell,cmd-in-legacy-dly",
+			     &slot->cmd_in_taps_dly[MMC_TIMING_LEGACY]);
+	of_property_read_u32(node, "marvell,data-in-legacy-dly",
+			     &slot->data_in_taps_dly[MMC_TIMING_LEGACY]);
 
 	max_frequency = max_supported_frequency(slot->host);
 
-- 
2.31.1

